### ISSUE TITLE
ADE20K dataset structure for semantic segmentation

### ISSUE THREAD

**Issue by pkwagner** [EPFL-VILAB/MultiMAE#19]
Hi,

First of all, thanks for your amazing work!

We're trying to reproduce the paper results and stumbled over how to set up semantic segmentation finetuning with ADE20K. To our surprise, the data loader seems to expect the same `root/task_a/class_x/xxx.ext` folder hierarchy known from classification. However, as the images naturally contain more than a single semantic class, we're not sure how the images are supposed to be arranged.

Could you hence give us a hint on how the data should be structured to work with the provided `ft_ade_64e_multimae-b_rgb.yaml` configuration?

Thank you,
Paul

**Comment by dmizr** [comment]
Hi @pkwagner ,

For tasks like semantic segmentation that don't involve classification, you can replace the `class_x` folder with a generic folder (e.g., one named `all` that contains all of the images). The dataloader is designed to work with a data structure where the folder names and filenames are consistent across tasks / modalities. You can find more information in the set-up section of the documentation: https://github.com/EPFL-VILAB/MultiMAE/blob/main/SETUP.md#dataset-structure

For ADE, the data is structured as such:
```
/path/to/ade/
├── train/
│   ├── rgb/
│   │   └── all/
│   │       ├── ADE_train_00000001.jpg
│   │       ├── ADE_train_00000002.jpg
│   │       └── ...
│   └── semseg/
│       └── all/
│           ├── ADE_train_00000001.png
│           ├── ADE_train_00000002.png
│           └── ...
└── val/
    ├── rgb/
    │   └── all/
    │       ├── ADE_val_00000001.jpg
    │       ├── ADE_val_00000002.jpg
    │       └── ...
    └── semseg/
        └── all/
            ├── ADE_val_00000001.png
            ├── ADE_val_00000002.png
            └── ...
```
where the `.png` files in the semseg subfolders contain the ground-truth semantic segmentation maps.

Best, David

**Comment by pkwagner** [comment]
Thanks!

Turns out we've used the wrong (full) dataset, which contains more than `COCO_SEMSEG_NUM_CLASSES = 133` labels and hence raised an assertion error.

We were a bit confused since the full dataset is also the one linked in `SETUP.md`. Assuming that the Scene Parsing Challenge dataset is indeed the correct one, wouldn't it make more sense to link to http://sceneparsing.csail.mit.edu instead of https://groups.csail.mit.edu/vision/datasets/ADE20K?

**Comment by dmizr** [comment]
Hi @pkwagner ,

When fine-tuning on ADE20K, we do use the full dataset with all 150 classes. You can find the name of the classes here: https://github.com/EPFL-VILAB/MultiMAE/blob/main/utils/datasets_semseg.py#L184

Please keep in mind that we only use RGB as the input modality when fine-tuning on ADE, so you won't need the `COCO_SEMSEG_NUM_CLASSES` constant unless you're using semantic segmentation as input (e.g., using a pseudo labeled COCO semseg map). If you're using our default config, which specifies `in_domains: rgb`, then you shouldn't encounter this error.

If you could provide the config that you're using, I would be happy to take a look and see if there's something we missed.

Best, David

**Comment by pkwagner** [comment]
Hm, that's strange.

We've used [this script [1]](https://termbin.com/tps7) to divide rgb and semseg to different paths (`{training,validation}/{rgb,semseg}/all`), with the resulting file tree looking like [this [2]](https://termbin.com/l89a). When using `ft_ade_64e_multimae-b_rgb.yaml` (only changing `finetune`, `data_path`, and `eval_data_path`) [[3]](https://termbin.com/npxos) with the full dataset, we get following error repeated over and over:

```
/opt/conda/conda-bld/pytorch_1634272068694/work/aten/src/ATen/native/cuda/NLLLoss2d.cu:95: nll_loss2d_forward_kernel: block: [4,0,0], thread: [590,0,0] Assertion `t >= 0 && t < n_classes` failed.
```

Digging a bit deeper into the code reveals that `seg_gt` indeed contains bigger values than allowed by `COCO_SEMSEG_NUM_CLASSES = 133`:

```python3
print(sorted(set(x.item() for x in seg_gt.flatten())))
# [16, 17, 22, 23, 27, 28, 29, 30, 33, 34, 35, 36, 39, 40, 45, 46, 47, 48, 51, 52, 53, 54, 58, 59, 63, 64, 65, 66, 69, 70, 71, 72, 75, 76, 77, 78, 81, 82, 83, 84, 87, 88, 89, 90, 93, 94, 95, 99, 100, 101, 105, 106, 107, 108, 111, 113, 114, 117, 118, 119, 123, 124, 125, 129, 130, 131, 135, 136, 137, 141, 142, 143, 153, 154, 155, 159, 160, 161, 165, 166, 167, 171, 172, 173, 177, 178, 179, 183, 184, 190, 191, 195, 196, 197, 201, 202, 203, 207, 208, 209, 213, 214, 215, 219, 220, 255]
```

When artificially limiting the value range with `seg_gt[seg_gt > COCO_SEMSEG_NUM_CLASSES - 1] = COCO_SEMSEG_NUM_CLASSES - 1`, everything works fine.

Since it's the only possibly variable part of the setup, I'm pretty sure it has to have to do with the dataset. Can you confirm that it is indeed the one provided via the link on the [ADE20K "Account Home" page](https://groups.csail.mit.edu/vision/datasets/ADE20K/request_data/account.php)?

**Comment by pkwagner** [comment]
Great, thanks a lot for helping out!
